### PR TITLE
Creating bigtable-hbase-beam

### PR DESCRIPTION
**Pull request by sduskis**
There are quite a few changes between the Beam and Dataflow connector:

1. `CloudBigtableOptions` were removed.  We may need to add it back for templatization
2. Added `CBTIOCoderRegistrar` to automatically register coders.
3. Removed the `Result[]` reader.
4. Depending on bigtable-hbase-1.x instead of bigtable-hbase-1.x-shaded

### CODE REVIEW

**Comment by sduskis** [comment]
This works, and is ready for review.

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/BigtableIOHBase.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+      this.serializableConfiguration =
+          new SerializableConfiguration(BigtableConfiguration.configure(projectId, instanceId));
+      this.tableId = tableId;
+    }
+
+    /**
+     * Add an extended configuration option.  See {@link BigtableOptionsFactory} for more details.
+     */
+    public Write withConfiguration(String key, String value) {
+      serializableConfiguration.get().set(key, value);
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/BigtableIOHBase.java]: same as above about immutability

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/BigtableIOHBase.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+          BigtableOptionsFactory.fromConfiguration(serializableConfiguration.get());
+      BigtableIO.Read bigtableRead =
+          BigtableIO.read()
+              .withBigtableOptions(options)
+              .withKeyRange(keyRange)
+              .withRowFilter(filter)
+              .withTableId(tableId);
+      return input.getPipeline()
+          .apply("BigtableIO.Read", bigtableRead)
+          .apply("HBase Result transformer", ParDo.of(ROW_TO_RESULT_TRANSFORM));
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/BigtableIOHBase.java]: Would be good to keep the naming consistent. Either CloudBigtableIO.transform or Bigtable read

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/BigtableIOHBase.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
+package com.google.cloud.bigtable.beam;
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/BigtableIOHBase.java]: Copyright

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/coders/HBaseMutationCoder.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/coders/HBaseMutationCoder.java]: Please update the year

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/coders/HBaseResultCoder.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/coders/HBaseResultCoder.java]: Please update the year

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/package-info.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright (C) 2015 Google Inc.
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/package-info.java]: Please update the year

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/test/java/com/google/cloud/bigtable/beam/coders/CoderTestUtil.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (C) 2015 Google Inc.
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/test/java/com/google/cloud/bigtable/beam/coders/CoderTestUtil.java]: Please update the year

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/test/java/com/google/cloud/bigtable/beam/coders/HBaseMutationCoderTest.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C) 2015 Google Inc.
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/test/java/com/google/cloud/bigtable/beam/coders/HBaseMutationCoderTest.java]: Please update the year

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/test/java/com/google/cloud/bigtable/beam/coders/HBaseResultCoderTest.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (C) 2015 Google Inc.
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/test/java/com/google/cloud/bigtable/beam/coders/HBaseResultCoderTest.java]: Please update the year

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/BigtableIOHBase.java [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+import com.google.cloud.bigtable.hbase.adapters.read.DefaultReadHooks;
+import com.google.cloud.bigtable.util.ZeroCopyByteStringUtil;
+import com.google.common.base.Preconditions;
+import com.google.protobuf.ByteString;
+
+/**
+ * Utilities to configure {@link BigtableIO} with HBase artifacts
+ */
+@Experimental
+public class BigtableIOHBase {
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/BigtableIOHBase.java]: I think BigtableHBaseIO would be more consistent with beam's naming convention

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/BigtableIOHBase.java [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+      this.serializableConfiguration =
+          new SerializableConfiguration(BigtableConfiguration.configure(projectId, instanceId));
+      this.tableId = tableId;
+    }
+
+    /**
+     * Add an extended configuration option.  See {@link BigtableOptionsFactory} for more details.
+     */
+    public Read withConfiguration(String key, String value) {
+      serializableConfiguration.get().set(key, value);
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/BigtableIOHBase.java]: Styleguide suggests similar,
Here is a rough sketch of using AutoValue to make this more consistent with Beam
https://github.com/sduskis/cloud-bigtable-client/pull/1

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/coders/CBTIOCoderRegistrar.java [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.apache.hadoop.hbase.client.Delete;
+import org.apache.hadoop.hbase.client.Mutation;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.Result;
+
+/**
+ * A {@link CoderProviderRegistrar} for standard types used with {@link BigtableIOHBase}.
+ */
+// TODO: remove this once HBaseIO's versions of these coders are public rather than package private.
```
Comment (sduskis) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/coders/CBTIOCoderRegistrar.java]: HBaseCoderProviderRegistrar does not exist in the 2.0.0 release.  It was added afterward.

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/BigtableIOHBase.java (hunk 10 above)
Comment (sduskis) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/BigtableIOHBase.java]: BigtableHBaseIO?  Do you have a suggestion?

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/coders/HBaseMutationCoder.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+ * Deletes to be processed on another Dataflow machine. This class handles the serialization and
+ * deserialization of the Puts and Deletes.
+ *
+ * <p> See CloudBigtableIO#initializeForWrite(Pipeline).
+ */
+public class HBaseMutationCoder extends AtomicCoder<Mutation> {
+
+  private static final long serialVersionUID = -3853654063196018580L;
+
+  public static Coder<?> of() {
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/coders/HBaseMutationCoder.java]: I think this should be Coder<Mutation>

File: bigtable-dataflow-parent/bigtable-hbase-beam/README.md [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+  // Dataflow properties
+  String zoneId = "Enter a zone Id, preferably the zone of your Cloud Bigtable cluster";
+  String dataflowGcsBucket = "gs://enter-your-bucket/and-optional-directory";
+
+  DataflowPipelineOptions options = PipelineOptionsFactory.as(DataflowPipelineOptions.class);
+  options.setProject(projectId);
+  options.setZone(zoneId);
+  options.setStagingLocation(dataflowGcsBucket + "/stage");
+  options.setTempLocation(dataflowGcsBucket + "/temp");
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/README.md]: I dont think tempLocation is necessary

File: bigtable-dataflow-parent/bigtable-hbase-beam/README.md [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+1. Add the words "Hello" and "World" to the pipeline.
+2. Converts the words to HBase Put objects.
+3. Sends the Puts to Cloud Bigtable.
+
+```java
+// Create a DoFn that creates a Put or Delete.  MUTATION_TRANSFORM is a simplistic example.
+static final DoFn<String, Mutation> MUTATION_TRANSFORM = new DoFn<String, Mutation>() {
+  @Override
+  public void processElement(DoFn<String, Mutation>.ProcessContext c) throws Exception {
+    c.output(new Put(c.element().getBytes()).addColumn(FAMILY, QUALIFIER, VALUE));
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/README.md]: s/VALUE/s.element()

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/coders/CBTIOCoderRegistrar.java [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+@AutoService(CoderProviderRegistrar.class)
+public class CBTIOCoderRegistrar implements CoderProviderRegistrar  {
+
+  @Override
+  public List<CoderProvider> getCoderProviders() {
+    return ImmutableList.of(
+      CoderProviders.forCoder(TypeDescriptor.of(Delete.class), HBaseMutationCoder.of()),
+      CoderProviders.forCoder(TypeDescriptor.of(Mutation.class), HBaseMutationCoder.of()),
+      CoderProviders.forCoder(TypeDescriptor.of(Put.class), HBaseMutationCoder.of()),
+      CoderProviders.forCoder(TypeDescriptor.of(Result.class), HBaseResultCoder.of()));
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/coders/CBTIOCoderRegistrar.java]: Is it necessary to specify Delete & Put when Mutation is already registered?

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/coders/HBaseMutationCoder.java [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.hadoop.hbase.client.Delete;
+import org.apache.hadoop.hbase.client.Mutation;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.protobuf.ProtobufUtil;
+import org.apache.hadoop.hbase.protobuf.generated.ClientProtos.MutationProto;
+import org.apache.hadoop.hbase.protobuf.generated.ClientProtos.MutationProto.MutationType;
+
+/**
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/coders/HBaseMutationCoder.java]: Please a note & a todo that this was copied from beam & should migrate to that version. Might be a good idea to make this package protected

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/coders/HBaseResultCoder.java [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.protobuf.ProtobufUtil;
+import org.apache.hadoop.hbase.protobuf.generated.ClientProtos;
+
+/**
+ * An {@link AtomicCoder} that serializes and deserializes the {@link Result}.
+ */
+public class HBaseResultCoder extends AtomicCoder<Result> {
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/coders/HBaseResultCoder.java]: similar comments as HBaseMutationCoder

**Comment by sduskis** [comment]
This should work towards fixing #1009

**Comment by reuvenlax** [comment]
What's the current status of this PR?

**Comment by sduskis** [comment]
This PR probably can't be merged due to some technical complexities.  We'll need a different solution for a Beam HBase Bigtable connector, specifically #1476.

**Comment by sduskis** [comment]
I'm going to close this for now.

**Pull request closed without merging**